### ISSUE TITLE
Document Proc failures in sink context, and other things in sink context too

### ISSUE THREAD

**Issue by briandfoy** [perl6/doc#1223]
- [x] Decide if its "void" or "sink" context. It's not like most Perlers even know there's a "void" context so referring to it doesn't help most people. Or, always say void context.
- [x] Document void context, probably next to other contexts, possibly in a document of basic syntax (duplicate of #732). The word "context" only shows up three times in [syntax](https://docs.perl6.org/language/syntax).
- [x] Document when a Proc object in sink context will throw an exception, which is only after its run and it thinks the exit code was nonzero
- [x] Same for Proc::Async
- [x] Fix some ambiguous wording in the FAQ that can seem to say the opposite of Failure.
- [x] Add sink and words related to it to the glossary
- [x] [Block](https://docs.perl6.org/type/Block) mentions that bare blocks in sink context execute immediately. 

In [RT 130715](https://rt.perl.org/Ticket/Display.html?id=130715), I report a problem with a Proc object throwing an exception. Since the last thing I did in the example was a call to `.close`, I attributed it to that action. It appears however, that doing nothing with the return value sinks it. That behavior is documented for the method sink, but that on its own implies it happens through affirmative action instead of implicitly.

However, the behavior of a Proc object in sink context isn't documented (perhaps #732 relates to this) that a Proc object will do this. There's a `.sink` method that says it will do this, but it doesn't say it's automatically called. Indeed, the only literal match for "sink" in docs.perl6.org search menu leads to Proc. It's also mentioned in [traps](https://docs.perl6.org/language/traps), but it uses the active "sinking" when this is really a problem of lack of action. 

[Proc::Async](https://docs.perl6.org/type/Proc$COLON$COLONAsync) mentions sink, but it doesn't say that not handling the result of await is from the sink context. By a stretch, if you know that was a thing, you'd get there eventually. But, almost no one knows that's a thing.

[5to6-perlfunc](https://docs.perl6.org/language/5to6-perlfunc) says that a `.sink` will be called when an object finds itself in "void context". As a nit, a Perl 6 object would never be in a context the language doesn't have. This very important tidbit about the language needs to be discussed on its own merits in a document that focuses on general syntax instead of particular classes. This relates to the glib statement I make in Perl classes: Python has poor docs organized well while Perl (5) has great documentation organized poorly. 

[faq](https://docs.perl6.org/language/faq) mentions sink, but only in relationship to Failure objects and "sinking" an object does the opposite of what I've shown so far. It appears to say it discards the error because there's a problem with parallel structure and pronouns in that sentence. The Failure docs are direct "Sink (void) context causes a Failure to throw, i.e. turn into a normal exception." Proc is not so direct, and neither is Proc::Async.

[Block](https://docs.perl6.org/type/Block) mentions that bare blocks in sink context execute immediately.

**Comment by AlexDaniel** [comment]
That's a nice lengthy bug report. FWIW, consider committing some changes directly (sometimes it's easier to change the damn thing instead of explaining what and why).

> However, the behavior of a Proc object in sink context isn't documented […]. There's a .sink method that says it will do this, but it doesn't say it's automatically called

Well, it says this:

> When sunk, the Proc object will throw X::Proc::Unsuccessful if the process it ran exited unsuccessfully.

I guess a couple of links can help: to whatever is produced for #732 and another link to [traps](https://docs.perl6.org/language/traps#Sunk_Proc).

And some consistency between Proc and Proc::Async would be nice, yes.

**Comment by b2gills** [comment]
There is a good reason why it is called `sink` context rather than `void` context.

That in part is because the word *void* is usually associated with the word *nothing*.  
Is the *nothing* about doing no actual work, or is it about throwing the result in `/dev/null`?

In Perl 5 for example something in `void` context always runs the associated code then throws away the result (barring any optimizations like using `WANTARRAY`).

In Haskell it wouldn't even run the associated code. (as far as I know)

I would guess that in most procedural languages `void` context is only about throwing the results away.  
In functional languages it is all about not even doing the work in the first place.

---

In Perl 6, `sink` context is about doing side effects only, and if there is any result throw it away.  
That is it tries to straddle the line between functional and procedural.

For example a Range `0..100` in `sink` context does absolutely nothing.

An Iterable/Iterator can do nothing if it knows it has no side effects.

A `map` Iterator for example, can't always know if its code object has side effects, so it has to run it.  
It does know that it doesn't have to keep the result because it is in `sink` context.  
(It's possible it could check if it was given a code object that has the `is pure` trait and do nothing)

A Failure that isn't handled always throws. If it is in `sink` context, it most definitely isn't handled.  
It can be passed all over your codebase, and doesn't throw until it gets used (a method call on it for example), or placed in `sink` context.

An object could run a different algorithm that doesn't keep values around that it otherwise would have, because it knows nothing is going to see them on the outside.

**Comment by JJ** [comment]
Well. Let's see what I can do with this.

**Comment by JJ** [comment]
Proc::Async.new does not throw in sink context. It might throw when you read from it, but in order to read from it you need it to *not* be in sink context. I mark that as solved.
